src/main/java/org/candlepin/model/AbstractHibernateCurator.java

```diff
@@ -996,24 +615,32 @@ public Collection<E> updateAll(Collection<E> entities, boolean flush, boolean ev
         return entities;
     }
 
+    private boolean isEntityNew(E entity) {
+        return entity.getId() == null;
+    }
+
     public Iterable<E> saveOrUpdateAll(Iterable<E> entities, boolean flush, boolean evict) {
         if (entities != null) {
             try {
-                Session session = this.currentSession();
                 EntityManager em = this.getEntityManager();
                 Iterable<List<E>> blocks = Iterables.partition(entities, getBatchBlockSize());
 
                 for (List<E> block : blocks) {
                     for (E entity : block) {
-                        session.saveOrUpdate(entity);
+                        if (isEntityNew(entity)) {
+                            em.persist(entity);
+                        }
+                        else {
+                            em.merge(entity);
+                        }
                     }
 
                     if (flush) {
                         em.flush();
 
                         if (evict) {
                             for (E entity : block) {
-                                session.evict(entity);
+                                em.detach(entity);
```

Comment:
I think that this should still be `session.evict` for the same reason that I left on the comment for AbstractHibernateCurator.evict. @Ceiu  am I understanding the difference between `Session.evict` and `EntityManager.detach` correctly?

Reply:
Inspecting the decompiled code that implements EntityManager.detach() (implemented by Hibernate's SessionImpl class) it looks like the detach method is just calling evict underneath. They're doing the same thing

Reply:
Okay perfect! Good idea on decompiling the code.